### ISSUE TITLE
suspend,hibernate: pam_systemd() call in runuser times out in sleep hooks, causing delay

### ISSUE THREAD

**Issue by rothn** [systemd/systemd#33020]
### systemd version the issue has been seen with

255.6-2

### Used distribution

Arch

### Linux kernel version used

6.9.1-arch1-2

### CPU architectures issue was seen on

None

### Expected behaviour you didn't see

`systemd suspend-then-hibernate` should wait to hibernate until battery is low, and sleep first.

### Unexpected behaviour you saw

`systemd suspend-then-hibernate` immediately hibernates, both on and off of AC power, on my Framework 16 laptop with dGPU.

### Steps to reproduce the problem

Run `systemd suspend-then-hibernate` and observe the issue.

### Additional program output to the terminal or log subsystem illustrating the issue

```sh
sleep.conf
---

[Sleep]
#AllowSuspend=yes
#AllowHibernation=yes
#AllowSuspendThenHibernate=yes
#AllowHybridSleep=yes
#SuspendState=mem standby freeze
#HibernateMode=platform shutdown
HybridSleepMode=suspend platform shutdown
HibernateDelaySec=3600
#SuspendEstimationSec=60min
```

logind.conf
---
```

[Login]
#NAutoVTs=6
#ReserveVT=6
#KillUserProcesses=no
#KillOnlyUsers=
#KillExcludeUsers=root
#InhibitDelayMaxSec=5
#UserStopDelaySec=10
#HandlePowerKey=poweroff
#HandlePowerKeyLongPress=ignore
#HandleRebootKey=reboot
#HandleRebootKeyLongPress=poweroff
#HandleSuspendKey=suspend
#HandleSuspendKeyLongPress=hibernate
#HandleHibernateKey=hibernate
#HandleHibernateKeyLongPress=ignore
HandleLidSwitch=suspend-then-hibernate
HandleLidSwitchExternalPower=suspend-then-hibernate
#HandleLidSwitchDocked=ignore
#PowerKeyIgnoreInhibited=no
#SuspendKeyIgnoreInhibited=no
#HibernateKeyIgnoreInhibited=no
#LidSwitchIgnoreInhibited=yes
#RebootKeyIgnoreInhibited=no
#HoldoffTimeoutSec=30s
#IdleAction=ignore
#IdleActionSec=30min
#RuntimeDirectorySize=10%
#RuntimeDirectoryInodesMax=
#RemoveIPC=yes
#InhibitorsMax=8192
#SessionsMax=8192
#StopIdleSessionSec=infinity
```
```

**Comment by rothn** [comment]
[it_always_hibernates.journalctl.log](https://github.com/systemd/systemd/files/15445204/it_always_hibernates.journalctl.log)

Attaching a full journalctl log of the issue. Note that while the log includes messages related to suspending, I know it hibernated because it takes a few minutes to come out of hibernation whereas coming out of a `systemd suspend` takes only a few seconds.

Additional context: Originally, this worked fine. Starting sometime around 255.1, I started noticing the issue and hoped it would go away with updates. It has not.

**Comment by poettering** [comment]
Hmm, that log only shows the system suspending "s2-idle", not hibernating?

**Comment by poettering** [comment]
> Note that while the log includes messages related to suspending, I know it hibernated because it takes a few minutes to come out of hibernation whereas coming out of a `systemd suspend` takes only a few seconds.

Uh, that's a wild over-interpretation. I figure instead just something is slowing down your regular suspend cycles?

Are things quick if you fire "systemctl suspend"?

**Comment by rothn** [comment]
> Are things quick if you fire "systemctl suspend"?

Yep, that works great-- the computer just goes to sleep! Ideally, that's what would happen when I close the lid instead of what currently does.

What I want is for `systemctl suspend` to happen when I close my lid and then `systemctl hibernate` to happen only when the computer is off for awhile.

> Uh, that's a wild over-interpretation. I figure instead just something is slowing down your regular suspend cycles?

I also know it's hibernating because when I open the computer, I see "Welcome to Arch Linux" and it decrypts my LUKS partition, mounts LVM, etc. I figure I wouldn't see that if it was just going to sleep. In that case, I'd expect behavior more like `systemctl suspend` where the computer is almost instantly usable.

I should have been more specific, and thanks for asking for clarification.

**Comment by rothn** [comment]
Update: The issue happens in both suspend and hibernate. @poettering , I owe you an apology. I accidentally capture an instance of the bug happening during suspend here. What was confusing me is that it affects both.

It looks like I might be hitting the issue in https://github.com/systemd/systemd/issues/2863 , but that was closed as completed in October 2017. Something delays the system by exactly 30 seconds when coming out of suspend or hibernate. My best guess right now is that this is the dbus delay mentioned in that issue, and perhaps related to a networking issue, since my wifi is always disconnected when I come out of a suspend or hibernate.

**So, is it possible that networking somehow delays my resume from suspend and hibernate by 2 minutes, 29 seconds each time? If so, is this a systemd bug? How would I be able to tell?** Happy to compile my own version with some debug stuff but I don't think I have time to try a bunch of different versions and bisect.


```
...
Jun 08 18:46:39 nickadeck systemd-logind[620]: Lid opened.
...
Jun 08 18:47:09 nickadeck runuser[35201]: pam_systemd(runuser-l:session): Failed to create session: Connection timed out
Jun 08 18:47:09 nickadeck runuser[35201]: pam_unix(runuser-l:session): session closed for user root
Jun 08 18:48:09 nickadeck syncthing[790]: [SXTJO] INFO: Lost primary connection to IVJUL6C at [fd94:2397:29a4:1de3:513c:cbb2:889:3696]:22000-[fd94:2397:29a4:1de3:c20c:92c9:a02b:9cce]:22000/tcp-server/TLS1.3-TLS_AES_128_GCM_SHA256/LAN-P10-5UN46UEI3AAOI98Q1LD0LRI0T0: read timeout (0 remain)
...
Jun 08 18:48:10 nickadeck systemd[35317]: Startup finished in 476ms.
```

**Comment by rothn** [comment]
Updated the ticket to reflect that disabling sleep hooks appears to be a workaround, and the issue appears to be a timeout in `pam_systemd` when `runuser` uses it to create a session. In my particular case (detailed at https://bbs.archlinux.org/viewtopic.php?pid=2177336#p2177336), the hook is `fw-fanctrl-suspend`:
```
#!/bin/sh

case $1 in
    pre)  runuser -l $(logname) -c "fw-fanctrl sleep" ;;
    post) runuser -l $(logname) -c "fw-fanctrl defaultStrategy" ;;
esac
```

**Comment by rothn** [comment]
This appears to be an issue reading from "/run/dbus/system_bus_socket"-- this times out, and eventually `runuser` continues without it. Is dbus supposed to be available to systemd sleep hooks?

**Comment by rothn** [comment]
This has now been discussed extensively at https://bbs.archlinux.org/viewtopic.php?pid=2178108#p2178108. See there for strace logs, dbus debug outputs, and other artifacts. It seems like this is an issue with `pam_systemd`.
